process/block/preprocess/transactions.go

```diff
@@ -948,11 +968,18 @@ func (txs *transactions) createAndProcessMiniBlocksFromMe(
 
 		senderAddressToSkip = []byte("")
 
-		gasRefunded := txs.gasHandler.GasRefunded(txHash)
 		if senderShardID == receiverShardID {
-			gasConsumedByMiniBlocksInSenderShard -= gasRefunded
-			totalGasConsumedInSelfShard -= gasRefunded
-			mapGasConsumedByMiniBlockInReceiverShard[receiverShardID] -= gasRefunded
+			gasRefunded := txs.gasHandler.GasRefunded(txHash)
+			gasPenalized := txs.gasHandler.GasPenalized(txHash)
+			gasToBeSubtracted := gasRefunded + gasPenalized
+			shouldDoTheSubtraction := gasToBeSubtracted <= gasConsumedByMiniBlocksInSenderShard &&
+				gasToBeSubtracted <= totalGasConsumedInSelfShard &&
+				gasToBeSubtracted <= mapGasConsumedByMiniBlockInReceiverShard[receiverShardID]
+			if shouldDoTheSubtraction {
+				gasConsumedByMiniBlocksInSenderShard -= gasToBeSubtracted
```

Comment:
same for other similar check in the other components.

Reply:
This could be used when we want to do this for only one variable. But here, and in the other cases you mentioned, we need to be sure that the condition is satisfied for more than one variable (2 or 3), and only if all of them satisfy the sanity check we will do the subtraction atomically from all of them. Otherwise, if the sanity check will fail for any of them you will cancel the subtraction for all of them. If I will apply the method you mentioned and let say the last subtraction will fail I should revert all the previous subtraction to the original values.